File: reference/merge-queue.ts
/**
 * SQLite-backed FIFO merge queue for agent branches.
 *
 * Backed by a SQLite database with WAL mode for concurrent access.
 * Uses bun:sqlite for zero-dependency, synchronous database access.
 * FIFO ordering guaranteed via autoincrement id.
 */

import { Database } from "bun:sqlite";
import { MergeError } from "../errors.ts";
import type { MergeEntry, ResolutionTier } from "../types.ts";

export interface MergeQueue {
	/** Add a new entry to the end of the queue with pending status. */
	enqueue(entry: Omit<MergeEntry, "enqueuedAt" | "status" | "resolvedTier">): MergeEntry;

	/** Remove and return the first pending entry, or null if none. */
	dequeue(): MergeEntry | null;

	/** Return the first pending entry without removing it, or null if none. */
	peek(): MergeEntry | null;

	/** List entries, optionally filtered by status. */
	list(status?: MergeEntry["status"]): MergeEntry[];

	/** Update the status (and optional resolution tier) of an entry by branch name. */
	updateStatus(branchName: string, status: MergeEntry["status"], tier?: ResolutionTier): void;

	/** Close the database connection. */
	close(): void;
}

/** Row shape as stored in SQLite (snake_case columns). */
interface MergeQueueRow {
	id: number;
	branch_name: string;
	task_id: string;
	agent_name: string;
	files_modified: string; // JSON array stored as text
	enqueued_at: string;
	status: string;
	resolved_tier: string | null;
}

const CREATE_TABLE = `
CREATE TABLE IF NOT EXISTS merge_queue (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  branch_name TEXT NOT NULL,
  task_id TEXT NOT NULL,
  agent_name TEXT NOT NULL,
  files_modified TEXT NOT NULL DEFAULT '[]',
  enqueued_at TEXT NOT NULL DEFAULT (strftime('%Y-%m-%dT%H:%M:%f','now')),
  status TEXT NOT NULL DEFAULT 'pending'
    CHECK(status IN ('pending','merging','merged','conflict','failed')),
  resolved_tier TEXT
    CHECK(resolved_tier IS NULL OR resolved_tier IN ('clean-merge','auto-resolve','ai-resolve','reimagine'))
)`;

const CREATE_INDEXES = `
CREATE INDEX IF NOT EXISTS idx_merge_queue_status ON merge_queue(status);
CREATE INDEX IF NOT EXISTS idx_merge_queue_branch ON merge_queue(branch_name)`;

/** Convert a database row (snake_case) to a MergeEntry object (camelCase). */
function rowToEntry(row: MergeQueueRow): MergeEntry {
	// Parse files_modified from JSON string to array, with fallback to empty array
	let filesModified: string[] = [];
	try {
		const parsed = JSON.parse(row.files_modified);
		filesModified = Array.isArray(parsed) ? parsed : [];
	} catch {
		// Fallback to empty array on parse error
		filesModified = [];
	}

	return {
		branchName: row.branch_name,
		taskId: row.task_id,
		agentName: row.agent_name,
		filesModified,
		enqueuedAt: row.enqueued_at,
		status: row.status as MergeEntry["status"],
		resolvedTier: row.resolved_tier as ResolutionTier | null,
	};
}

/**
 * Migrate an existing merge_queue table from bead_id to task_id column.
 * Safe to call multiple times — only renames if bead_id exists and task_id does not.
 */
function migrateBeadIdToTaskId(db: Database): void {
	const rows = db.prepare("PRAGMA table_info(merge_queue)").all() as Array<{ name: string }>;
	const existingColumns = new Set(rows.map((r) => r.name));
	if (existingColumns.has("bead_id") && !existingColumns.has("task_id")) {
		db.exec("ALTER TABLE merge_queue RENAME COLUMN bead_id TO task_id");
	}
}

/**
 * Create a new MergeQueue backed by a SQLite database at the given path.
 *
 * Initializes the database with WAL mode and a 5-second busy timeout.
 * Creates the merge_queue table and indexes if they do not already exist.
 */
export function createMergeQueue(dbPath: string): MergeQueue {
	const db = new Database(dbPath);

	// Configure for concurrent access from multiple agent processes
	db.exec("PRAGMA journal_mode = WAL");
	db.exec("PRAGMA synchronous = NORMAL");
	db.exec("PRAGMA busy_timeout = 5000");

	// Create schema
	db.exec(CREATE_TABLE);
	db.exec(CREATE_INDEXES);

	// Migrate: rename bead_id → task_id on existing tables
	migrateBeadIdToTaskId(db);

	// Prepare statements for frequent operations
	const insertStmt = db.prepare<
		MergeQueueRow,
		{
			$branch_name: string;
			$task_id: string;
			$agent_name: string;
			$files_modified: string;
			$enqueued_at: string;
		}
	>(`
		INSERT INTO merge_queue (branch_name, task_id, agent_name, files_modified, enqueued_at)
		VALUES ($branch_name, $task_id, $agent_name, $files_modified, $enqueued_at)
		RETURNING *
	`);

	const getFirstPendingStmt = db.prepare<MergeQueueRow, Record<string, never>>(`
		SELECT * FROM merge_queue WHERE status = 'pending' ORDER BY id ASC LIMIT 1
	`);

	const deleteByIdStmt = db.prepare<void, { $id: number }>(`
		DELETE FROM merge_queue WHERE id = $id
	`);

	const listAllStmt = db.prepare<MergeQueueRow, Record<string, never>>(`
		SELECT * FROM merge_queue ORDER BY id ASC
	`);

	const listByStatusStmt = db.prepare<MergeQueueRow, { $status: string }>(`
		SELECT * FROM merge_queue WHERE status = $status ORDER BY id ASC
	`);

	const getByBranchStmt = db.prepare<MergeQueueRow, { $branch_name: string }>(`
		SELECT * FROM merge_queue WHERE branch_name = $branch_name
	`);

	const updateStatusStmt = db.prepare<
		void,
		{
			$branch_name: string;
			$status: string;
			$resolved_tier: string | null;
		}
	>(`
		UPDATE merge_queue
		SET status = $status, resolved_tier = $resolved_tier
		WHERE branch_name = $branch_name
	`);

	return {
		enqueue(input): MergeEntry {
			const filesModifiedJson = JSON.stringify(input.filesModified);
			const enqueuedAt = new Date().toISOString();

			const row = insertStmt.get({
				$branch_name: input.branchName,
				$task_id: input.taskId,
				$agent_name: input.agentName,
				$files_modified: filesModifiedJson,
				$enqueued_at: enqueuedAt,
			});

			if (row === null || row === undefined) {
				throw new MergeError("Failed to insert entry into merge queue");
			}

			return rowToEntry(row);
		},

		dequeue(): MergeEntry | null {
			const row = getFirstPendingStmt.get({});

			if (row === null || row === undefined) {
				return null;
			}

			// Delete the entry
			deleteByIdStmt.run({ $id: row.id });

			return rowToEntry(row);
		},

		peek(): MergeEntry | null {
			const row = getFirstPendingStmt.get({});

			if (row === null || row === undefined) {
				return null;
			}

			return rowToEntry(row);
		},

		list(status?): MergeEntry[] {
			let rows: MergeQueueRow[];

			if (status === undefined) {
				rows = listAllStmt.all({});
			} else {
				rows = listByStatusStmt.all({ $status: status });
			}

			return rows.map(rowToEntry);
		},

		updateStatus(branchName, status, tier?): void {
			// Check if entry exists
			const existing = getByBranchStmt.get({ $branch_name: branchName });

			if (existing === null || existing === undefined) {
				throw new MergeError(`No queue entry found for branch: ${branchName}`, {
					branchName,
				});
			}

			// Update the entry
			updateStatusStmt.run({
				$branch_name: branchName,
				$status: status,
				$resolved_tier: tier ?? null,
			});
		},

		close(): void {
			db.exec("PRAGMA wal_checkpoint(PASSIVE)");
			db.close();
		},
	};
}

File: src/db/merge_queue.rs
#![allow(dead_code)]

use rusqlite::{params, Connection, OptionalExtension};

use crate::db::connection::open_connection;
use crate::errors::{GroveError, Result};
use crate::types::{InsertMergeEntry, MergeEntry, MergeEntryStatus, ResolutionTier};

const MERGE_QUEUE_SCHEMA: &str = "
CREATE TABLE IF NOT EXISTS merge_queue (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  branch_name TEXT NOT NULL,
  task_id TEXT NOT NULL,
  agent_name TEXT NOT NULL,
  files_modified TEXT NOT NULL DEFAULT '[]',
  enqueued_at TEXT NOT NULL DEFAULT (strftime('%Y-%m-%dT%H:%M:%f','now')),
  status TEXT NOT NULL DEFAULT 'pending'
    CHECK(status IN ('pending','merging','merged','conflict','failed')),
  resolved_tier TEXT
    CHECK(resolved_tier IS NULL OR resolved_tier IN ('clean-merge','auto-resolve','ai-resolve','reimagine'))
);
CREATE INDEX IF NOT EXISTS idx_merge_queue_status ON merge_queue(status);
CREATE INDEX IF NOT EXISTS idx_merge_queue_branch ON merge_queue(branch_name);
";

fn row_to_entry(row: &rusqlite::Row<'_>) -> rusqlite::Result<MergeEntry> {
    let files_json: String = row.get(4)?;
    let files_modified: Vec<String> = serde_json::from_str(&files_json).unwrap_or_default();
    Ok(MergeEntry {
        id: row.get(0)?,
        branch_name: row.get(1)?,
        task_id: row.get(2)?,
        agent_name: row.get(3)?,
        files_modified,
        enqueued_at: row.get(5)?,
        status: row.get(6)?,
        resolved_tier: row.get(7)?,
    })
}

pub struct MergeQueue {
    conn: Connection,
}

impl MergeQueue {
    pub fn new(path: &str) -> Result<Self> {
        let conn = open_connection(path)?;
        conn.execute_batch(MERGE_QUEUE_SCHEMA)?;
        Ok(Self { conn })
    }

    pub fn enqueue(&self, entry: &InsertMergeEntry) -> Result<MergeEntry> {
        let files_json = serde_json::to_string(&entry.files_modified)?;
        let now = chrono::Utc::now().to_rfc3339();
        self.conn.execute(
            "INSERT INTO merge_queue (branch_name, task_id, agent_name, files_modified, enqueued_at, status)
             VALUES (?1, ?2, ?3, ?4, ?5, 'pending')",
            params![entry.branch_name, entry.task_id, entry.agent_name, files_json, now],
        )?;
        let id = self.conn.last_insert_rowid();
        Ok(MergeEntry {
            id,
            branch_name: entry.branch_name.clone(),
            task_id: entry.task_id.clone(),
            agent_name: entry.agent_name.clone(),
            files_modified: entry.files_modified.clone(),
            enqueued_at: now,
            status: MergeEntryStatus::Pending,
            resolved_tier: None,
        })
    }

    pub fn dequeue(&mut self) -> Result<Option<MergeEntry>> {
        let tx = self.conn.transaction()?;
        let result = tx
            .query_row(
                "SELECT id, branch_name, task_id, agent_name, files_modified, enqueued_at, status, resolved_tier
                 FROM merge_queue WHERE status = 'pending' ORDER BY id ASC LIMIT 1",
                [],
                row_to_entry,
            )
            .optional()?;

        if let Some(ref entry) = result {
            tx.execute(
                "UPDATE merge_queue SET status = 'merging' WHERE id = ?1",
                params![entry.id],
            )?;
        }
        tx.commit()?;

        // Return with updated status if we found an entry
        Ok(result.map(|mut e| {
            e.status = MergeEntryStatus::Merging;
            e
        }))
    }

    pub fn peek(&self) -> Result<Option<MergeEntry>> {
        let result = self
            .conn
            .query_row(
                "SELECT id, branch_name, task_id, agent_name, files_modified, enqueued_at, status, resolved_tier
                 FROM merge_queue WHERE status = 'pending' ORDER BY id ASC LIMIT 1",
                [],
                row_to_entry,
            )
            .optional()?;
        Ok(result)
    }

    pub fn list(&self, status: Option<MergeEntryStatus>) -> Result<Vec<MergeEntry>> {
        let sql = if let Some(ref s) = status {
            let s_str = match s {
                MergeEntryStatus::Pending => "pending",
                MergeEntryStatus::Merging => "merging",
                MergeEntryStatus::Merged => "merged",
                MergeEntryStatus::Conflict => "conflict",
                MergeEntryStatus::Failed => "failed",
            };
            format!(
                "SELECT id, branch_name, task_id, agent_name, files_modified, enqueued_at, status, resolved_tier
                 FROM merge_queue WHERE status = '{}' ORDER BY id ASC",
                s_str
            )
        } else {
            String::from(
                "SELECT id, branch_name, task_id, agent_name, files_modified, enqueued_at, status, resolved_tier
                 FROM merge_queue ORDER BY id ASC",
            )
        };
        let mut stmt = self.conn.prepare(&sql)?;
        let rows = stmt.query_map([], row_to_entry)?;
        rows.collect::<rusqlite::Result<Vec<_>>>()
            .map_err(GroveError::from)
    }

    pub fn update_status(
        &self,
        branch_name: &str,
        status: MergeEntryStatus,
        tier: Option<ResolutionTier>,
    ) -> Result<()> {
        self.conn.execute(
            "UPDATE merge_queue SET status = ?1, resolved_tier = ?2 WHERE branch_name = ?3",
            params![status, tier, branch_name],
        )?;
        Ok(())
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    fn make_entry(branch: &str) -> InsertMergeEntry {
        InsertMergeEntry {
            branch_name: branch.to_string(),
            task_id: "task-001".to_string(),
            agent_name: "agent-a".to_string(),
            files_modified: vec!["src/main.rs".to_string(), "src/lib.rs".to_string()],
        }
    }

    #[test]
    fn test_schema_idempotent() {
        MergeQueue::new(":memory:").unwrap();
    }

    #[test]
    fn test_enqueue_and_list() {
        let store = MergeQueue::new(":memory:").unwrap();
        let entry = store.enqueue(&make_entry("feature-a")).unwrap();
        assert_eq!(entry.branch_name, "feature-a");
        assert_eq!(entry.status, MergeEntryStatus::Pending);
        assert_eq!(entry.files_modified, vec!["src/main.rs", "src/lib.rs"]);

        let list = store.list(None).unwrap();
        assert_eq!(list.len(), 1);
    }

    #[test]
    fn test_files_modified_roundtrip() {
        let store = MergeQueue::new(":memory:").unwrap();
        let entry = store.enqueue(&make_entry("branch-x")).unwrap();
        let list = store.list(None).unwrap();
        assert_eq!(list[0].files_modified, entry.files_modified);
    }

    #[test]
    fn test_peek() {
        let store = MergeQueue::new(":memory:").unwrap();
        assert!(store.peek().unwrap().is_none());
        store.enqueue(&make_entry("branch-a")).unwrap();
        let peeked = store.peek().unwrap().unwrap();
        assert_eq!(peeked.branch_name, "branch-a");
        // Peek doesn't change status
        let peeked2 = store.peek().unwrap().unwrap();
        assert_eq!(peeked2.status, MergeEntryStatus::Pending);
    }

    #[test]
    fn test_dequeue_fifo_order() {
        let mut store = MergeQueue::new(":memory:").unwrap();
        store.enqueue(&make_entry("branch-a")).unwrap();
        store.enqueue(&make_entry("branch-b")).unwrap();

        let first = store.dequeue().unwrap().unwrap();
        assert_eq!(first.branch_name, "branch-a");
        assert_eq!(first.status, MergeEntryStatus::Merging);

        let second = store.dequeue().unwrap().unwrap();
        assert_eq!(second.branch_name, "branch-b");
    }

    #[test]
    fn test_dequeue_empty() {
        let mut store = MergeQueue::new(":memory:").unwrap();
        assert!(store.dequeue().unwrap().is_none());
    }

    #[test]
    fn test_update_status() {
        let store = MergeQueue::new(":memory:").unwrap();
        store.enqueue(&make_entry("branch-a")).unwrap();
        store
            .update_status(
                "branch-a",
                MergeEntryStatus::Merged,
                Some(ResolutionTier::CleanMerge),
            )
            .unwrap();

        let list = store.list(Some(MergeEntryStatus::Merged)).unwrap();
        assert_eq!(list.len(), 1);
        assert_eq!(list[0].resolved_tier, Some(ResolutionTier::CleanMerge));
    }

    #[test]
    fn test_list_by_status() {
        let store = MergeQueue::new(":memory:").unwrap();
        store.enqueue(&make_entry("a")).unwrap();
        store.enqueue(&make_entry("b")).unwrap();
        store
            .update_status("a", MergeEntryStatus::Merged, None)
            .unwrap();

        let pending = store.list(Some(MergeEntryStatus::Pending)).unwrap();
        assert_eq!(pending.len(), 1);
        assert_eq!(pending[0].branch_name, "b");

        let merged = store.list(Some(MergeEntryStatus::Merged)).unwrap();
        assert_eq!(merged.len(), 1);
    }
}
